Added a packet defragmenter

blanu opened the pull request:
The packet defragmenter caches fragments of a fragmented packet until all fragments have arrived.
Fragments are evicted from the cache when all packets have arrived.
There is also a reaper for each packet that will evict fragments from the cache if all fragments for a packet have not arrived one minute after the arrival of the first fragment for that packet.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy-lib/274)

<!-- Reviewable:end -->

Commit:
Added a packet defragmenter

trevj commented:
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 7 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy8Ou2JANmLEsL5LgkH)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/9f29b8d4a9509fbc992736e4978100d3a80230f8/src/fancy-transformers/defragmenter.ts#L7)):</sup>
Are you wholly or mostly following RFC 815 here? Worth mentioning.

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 9 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy8N3I51YKQapQbkwfv)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/9f29b8d4a9509fbc992736e4978100d3a80230f8/src/fancy-transformers/defragmenter.ts#L9)):</sup>
No need to repeat the variable name in the comment, e.g.:
```typescript
// Associates packet identifiers with indexed lists of fragments.
private tracker_ :{[index:string]:ArrayBuffer[]} = {};
```

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 10 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy8NI87jVWKlC8UZ42b)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/9f29b8d4a9509fbc992736e4978100d3a80230f8/src/fancy-transformers/defragmenter.ts#L10)):</sup>
Add a comment describing the map keys: fragment IDs, in hex format.

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 11 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy8NDJnTmzA2IV3piij)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/9f29b8d4a9509fbc992736e4978100d3a80230f8/src/fancy-transformers/defragmenter.ts#L11)):</sup>
Spaces between field declarations!

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 58 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy8SP0meQApeLjGNKfq)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/9f29b8d4a9509fbc992736e4978100d3a80230f8/src/fancy-transformers/defragmenter.ts#L58)):</sup>
You have two maps indexed by the same key, which is a code smell.

Normally you'd create a struct to hold the data but in this case - because each fragment has a length field - I *think* you can remote `counter_` entirely:
```typescript
// If we have all fragments for this packet identifier, it is complete.
if (this.tracker_[hexid].length === this.tracker_[hexid][0].count) {
  this.complete_.push(hexid);
}
```

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 90 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy8Qdt9_7bq70INW7u1)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/9f29b8d4a9509fbc992736e4978100d3a80230f8/src/fancy-transformers/defragmenter.ts#L90)):</sup>
No magic numbers! Extract the 60 as a constant.

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 100 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy8OZhcFsPIx0qZgyXt)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/9f29b8d4a9509fbc992736e4978100d3a80230f8/src/fancy-transformers/defragmenter.ts#L100)):</sup>
This also clears `this.complete_`, right?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274)
<!-- Sent from Reviewable.io -->

Commit:
Added additional comment detail about use of cache expiration strategy

Commit:
Formatting changes to stay within line length limit from style guide

Commit:
Removed variable names from variable comments

Commit:
Rewrote defragmenter to use a map to a structure instead of parallel maps

Commit:
Replaced hardcoded cache expiration timeout with a const

Commit:
Blank newlines between field declarations

blanu commented:
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 7 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy8Ou2JANmLEsL5LgkH)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/9f29b8d4a9509fbc992736e4978100d3a80230f8/src/fancy-transformers/defragmenter.ts#L7)):</sup>
I am only using the cache expiration strategy from RFC 815. This the result of a discussion in which I asks for Ben's opinion on the cache expiration strategy and he said that implementation of a strategy from an RFC would be preferable to inventing a strategy from first principles. I have added a comment to clarify that it is just the cache expiration strategy that is being used from RFC 815.

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 9 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy8N3I51YKQapQbkwfv)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/9f29b8d4a9509fbc992736e4978100d3a80230f8/src/fancy-transformers/defragmenter.ts#L9)):</sup>
Okay I have made this change.

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 10 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy8NI87jVWKlC8UZ42b)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/9f29b8d4a9509fbc992736e4978100d3a80230f8/src/fancy-transformers/defragmenter.ts#L10)):</sup>
Okay I have made this change.

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 58 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy8SP0meQApeLjGNKfq)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/9f29b8d4a9509fbc992736e4978100d3a80230f8/src/fancy-transformers/defragmenter.ts#L58)):</sup>
Okay I have merged the parallel maps into a single map to a structure.

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 90 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy8Qdt9_7bq70INW7u1)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/9f29b8d4a9509fbc992736e4978100d3a80230f8/src/fancy-transformers/defragmenter.ts#L90)):</sup>
Okay I have replaced this with a const.

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 100 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy8OZhcFsPIx0qZgyXt)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/9f29b8d4a9509fbc992736e4978100d3a80230f8/src/fancy-transformers/defragmenter.ts#L100)):</sup>
Correct. I have added a comment to this effect.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274)
<!-- Sent from Reviewable.io -->

blanu commented:
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 11 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy8NDJnTmzA2IV3piij)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/9f29b8d4a9509fbc992736e4978100d3a80230f8/src/fancy-transformers/defragmenter.ts#L11)):</sup>
Okay I have made this change.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274)
<!-- Sent from Reviewable.io -->

trevj commented:
OK. There's still a little too much book-keeping for my taste - e.g. we could remove the counters if we used something like a heap, keyed by fragment number, to store the fragments - especially given the lack of tests, but this is fine for a first pass.

A couple of minor suggestions, then :+1:.

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 8 \[r2\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy9VIPuHcmQjFn453tE)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/a756d2eed417e1c43e31b008a14e814a1c449a6e/src/fancy-transformers/defragmenter.ts#L8)):</sup>
`const`! Cool.

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 124 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy8OZhcFsPIx0qZgyXt)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/9f29b8d4a9509fbc992736e4978100d3a80230f8/src/fancy-transformers/defragmenter.ts#L124)):</sup>
Clearing `this.complete_` is an implementation detail (put another way, you are referencing a private variable in a public method's comment); you should describe that this clears the set of stored completed fragments.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274)
<!-- Sent from Reviewable.io -->

Commit:
Modified public method comment to remove reference to private variable

blanu commented:
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

<sup>**[src/fancy-transformers/defragmenter.ts, line 100 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274#-Jy8OZhcFsPIx0qZgyXt)** ([raw file](https://github.com/uproxy/uproxy-lib/blob/9f29b8d4a9509fbc992736e4978100d3a80230f8/src/fancy-transformers/defragmenter.ts#L100)):</sup>
Okay I have made this change.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/uproxy-lib/274)
<!-- Sent from Reviewable.io -->
